Ref bitshares-core#1584 Fix some warnings

nathanielhourt opened the pull request.

nathanielhourt committed:
Fix some warnings

nathanielhourt commented on include/fc/exception/exception.hpp:
```diff
@@ -446,11 +446,11 @@ namespace fc
                 fc::std_exception_code,\
                 typeid(e).name(), \
-                e.what() ) ; throw fce;\
+                e.what() ) ;\
```
I get a warning on these called "throw by value, catch by reference" -- the compiler recommended throwing an anonymous value if you plan to catch by reference, since a named, stack-allocated value technically can't be caught by reference (the stack it was allocated on has been destroyed).

Compiler handles it correctly either way, but this way is more complicit with formal C++ memory rules.

nathanielhourt commented on tests/crypto/bigint_test.cpp:
```diff
@@ -46,8 +46,6 @@ BOOST_AUTO_TEST_CASE(bigint_test_2)
     } while ( bi_accu.log2() <= 128 );
 
-    bi_accu = bi_accu;
-
```
I checked that bigint does properly ignore a self-assignment (and self-move), and this line causes warnings. I think we can safely do away with it.

pmconrad replied:
IIRC I added this line on purpose, to test just that. Can you trick the compiler? Perhaps by assigning a reference to itself or sth?

nathanielhourt replied:
OK, so I've added it back but with enough nonsense to confuse the compiler into just doing it and not warning us about it. How's that?

nathanielhourt commented on tests/io/json_tests.cpp:
```diff
@@ -171,7 +171,7 @@ static bool equal( const fc::variant& a, const fc::variant& b )
          return true;
       case fc::variant::type_id::blob_type:
       default:
-         FC_THROW_EXCEPTION( fc::invalid_arg_exception, "Unsupported variant type: " + a.get_type() );
```
`char[] + int` doesn't make sense, but `char[] + string` does.

pmconrad replied:
Better(?) alternative: use format string and int argument. See https://github.com/bitshares/bitshares-fc/pull/102#pullrequestreview-189925457

nathanielhourt replied:
Huh, yeah that totally works. I just assumed it didn't since it wasn't there to begin with. :]

nathanielhourt committed:
Fix warning

Not sure what the story is with this file, but it's making a
warning and this fixes it.

nathanielhourt committed:
Add self-assignment test, sans the compiler warning

nathanielhourt committed:
More tweaks vis-a-vis warnings

Move crc_c fake usage out of strange main() function in a library
Convert manual string formatting to FC string formatting
